### PR TITLE
CPP SDK : changes to handle additional Properties

### PR DESCRIPTION
**Pull request by HaseenaSainul**

### CODE REVIEW

File: src/macrofier/types.mjs [hunk 1]
```diff
@@ -137,6 +139,39 @@ const insertEnumMacros = (content, schema, module, name) => {
   return template.join('\n')
 }
 
+const insertObjectAdditionalPropertiesMacros = (content, schema, module, title, options) => {
+  options = options ? JSON.parse(JSON.stringify(options)) : {}
+
+  const options2 = JSON.parse(JSON.stringify(options))
+  options2.parent = title
```
Comment (jlacivita) [inline src/macrofier/types.mjs:146]: You've already made a copy on line 143. Why do you need a second copy?

File: src/macrofier/types.mjs [hunk 2]
```diff
@@ -137,6 +139,39 @@ const insertEnumMacros = (content, schema, module, name) => {
...
+  let jsonType = getJsonType(schema.additionalProperties, module)
+  if (!isPrimitiveType(jsonType)) {
+    jsonType = 'string'
+  }
+
+  const additionalType = getPrimitiveType(jsonType, 'additional-types')
+  let key = (propertyNames && propertyNames.title) ? propertyNames.title : getTemplate(path.join(options.templateDir, 'additionalPropertiesKey')).trimEnd()
+
+  content = content
+    .replace(/\$\{shape\}/g, shape)
```
Comment (jlacivita) [inline src/macrofier/types.mjs:163]: Did you confirm that none of the JS templates were using `${property}` or `${Property}` here?

I forget why I had those two lines in there, but there must have been some reason.

File: src/macrofier/types.mjs [hunk 3]
```diff
@@ -27,9 +27,11 @@ const primitives = {
   "integer": "number",
   "number": "number",
   "boolean": "boolean",
-  "string": "string"
+  "string": "string",
+  "float": "float"
```
Comment (jlacivita) [inline src/macrofier/types.mjs:31]: The keys in the map are JSON-Schema primitives. JSON-schema does not have a `float` type, so it should not be here.

Did you come across a schema w/ a `float` type? I'm surprised it passed validation.

Comment (jlacivita) [reply src/macrofier/types.mjs:31]: https://json-schema.org/understanding-json-schema/reference/numeric#:~:text=JSON%20does%20not%20have%20distinct,the%20same%20value%20in%20JSON.

Comment (jlacivita) [reply src/macrofier/types.mjs:31]: These are just default values for JavaScript. If you need different overrides for Cpp, they can go in the sdk.config.json

File: src/macrofier/types.mjs (hunk 1 above)
Comment (jlacivita) [reply src/macrofier/types.mjs:146]: Also, I can't find where this new `insertObjectAdditionalProperitesMacros` method is called.

File: src/macrofier/types.mjs [hunk 4]
```diff
@@ -151,21 +186,7 @@ const insertObjectMacros = (content, schema, module, title, property, options) =
...
-        .replace(/\$\{property\}/g, '')
-        .replace(/\$\{Property\}/g, '')
-        .replace(/\$\{shape\}/g, shape)
-        .replace(/\$\{parent\.title\}/g, title)
-        .replace(/\$\{title\}/g, type)
-        .replace(/\$\{delimiter\}(.*?)\$\{end.delimiter\}/g, '')
-        .replace(/\$\{if\.optional\}(.*?)\$\{end\.if\.optional\}/g, ''))
-    }
-  
+
```
Comment (jlacivita) [inline src/macrofier/types.mjs:189]: Shouldn't we be calling  `insertObjectAdditionalProperitesMacros` here, with the same `if (schema.additionalProperties...` block?

File: src/macrofier/types.mjs [hunk 5]
```diff
@@ -614,11 +635,17 @@ function getSchemaType(schema, module, { destination, templateDir = 'types', lin
     // }
   }
   else if (schema.type) {
-    // TODO: this assumes that when type is an array of types, that it's one other primative & 'null', which isn't necessarily true.
-    const schemaType = !Array.isArray(schema.type) ? schema.type : schema.type.find(t => t !== 'null')
-    const primitive = getPrimitiveType(schemaType, templateDir)
-    const type = allocatedProxy ? allocatedPrimitiveProxies[schemaType] || primitive : primitive
-    return wrap(type, code ? '`' : '')
+    const template = getTemplate(path.join(templateDir, 'additionalProperties'))
```
Comment (jlacivita) [inline src/macrofier/types.mjs:638]: The current method (`getSchemaType`) doesn't recurse into each property... 

It looks like you're inserting the `Title` template if there happens to be a template for `additionalProperties`

I'm not sure I follow that. Could you explain?

File: src/macrofier/types.mjs [hunk 6]
```diff
@@ -644,4 +685,4 @@ export default {
   getSchemaType,
   getJsonType,
```
Comment (jlacivita) [inline src/macrofier/types.mjs:686]: getJSONType wasn't really being used before. I see that you're using it to do some local work in this method, which is cool, but we should stop exporting it now that you've got logic there.

Don't want any external dependencies on it.

**Review by jlacivita — CHANGES_REQUESTED** [review]

File: src/macrofier/types.mjs (hunk 5 above)
Comment (HaseenaSainul) [reply src/macrofier/types.mjs:638]: additionalPorperties always seening like a map. FlatMap is additional property and it is map. Here we have to get only its title. FlatMap. hence added logic to get Title.

File: src/macrofier/types.mjs (hunk 1 above)
Comment (HaseenaSainul) [reply src/macrofier/types.mjs:146]: sorry missed to updated calling place. corrected now

File: src/macrofier/types.mjs (hunk 2 above)
Comment (HaseenaSainul) [reply src/macrofier/types.mjs:163]: I have not seen any use case from javascript. javascript generation not showed any issue

File: src/macrofier/types.mjs (hunk 3 above)
Comment (HaseenaSainul) [reply src/macrofier/types.mjs:31]: yea, it is already overridden language config. I missed to remove. initially I did not clear idea about this setting. finally I understood and missed to remove it

File: src/macrofier/types.mjs (hunk 4 above)
Comment (HaseenaSainul) [reply src/macrofier/types.mjs:189]: I updated code now to call insertObjectAdditionalProperitesMacros now

File: src/macrofier/types.mjs (hunk 6 above)
Comment (HaseenaSainul) [reply src/macrofier/types.mjs:686]: removed it

**Pull request merged**